File: charsibot/commands.go
package charsibot

import (
	"context"
	"fmt"
	"strings"

	"github.com/joeyak/go-twitch-eventsub/v3"

	"github.com/lukeramljak/charsibot/blindbox"
	"github.com/lukeramljak/charsibot/server"
	"github.com/lukeramljak/charsibot/stats"
)

type Command struct {
	Execute func(ctx context.Context, b *Bot, event twitch.EventChannelChatMessage)
}

// Commands returns the full map of chat commands keyed by trigger word.
func Commands(seriesConfigs []blindbox.SeriesConfig) map[string]Command {
	cmds := map[string]Command{
		"collections": {
			Execute: func(ctx context.Context, b *Bot, _ twitch.EventChannelChatMessage) {
				collections, err := b.blindboxService.GetCompletedCollections(ctx)
				if err != nil {
					b.logger.Error("failed to get completed collections", "err", err)
					return
				}
				b.SendMessage(SendMessageParams{Message: "The following chatters have completed the below blind box collections:"})
				for _, row := range collections {
					b.SendMessage(SendMessageParams{Message: fmt.Sprintf("%s: %s", row.SeriesName, row.Usernames)})
				}
			},
		},
		"leaderboard": {
			Execute: func(ctx context.Context, b *Bot, _ twitch.EventChannelChatMessage) {
				rows, err := b.statsService.GetStatLeaderboard(ctx)
				if err != nil {
					b.logger.Error("failed to get leaderboard", "err", err)
					b.SendMessage(SendMessageParams{Message: "Failed to get leaderboard"})
					return
				}
				parts := make([]string, len(rows))
				for i, row := range rows {
					parts[i] = fmt.Sprintf("%s %s (%d)", row.Emoji, row.Username, row.Value)
				}
				b.SendMessage(SendMessageParams{Message: strings.Join(parts, " | ")})
			},
		},
		"stats": {
			Execute: func(ctx context.Context, b *Bot, event twitch.EventChannelChatMessage) {
				if len(strings.Fields(event.Message.Text)) != 1 {
					return
				}
				userStats, err := b.statsService.GetOrCreateStats(ctx, event.ChatterUserId, event.ChatterUserName)
				if err != nil {
					b.logger.Error("failed to get stats", "err", err, "user", event.ChatterUserName)
					return
				}
				b.SendMessage(SendMessageParams{Message: stats.FormatStats(event.ChatterUserName, userStats), ReplyParentMessageID: event.MessageId})
			},
		},
	}

	for _, cfg := range seriesConfigs {
		cmds[cfg.Series] = Command{
			Execute: func(ctx context.Context, b *Bot, event twitch.EventChannelChatMessage) {
				if len(strings.Fields(event.Message.Text)) != 1 {
					return
				}
				slots, err := b.blindboxService.GetCollection(ctx, event.ChatterUserId, cfg.Series)
				if err != nil {
					b.logger.Error("failed to get collection", "err", err, "user", event.ChatterUserName)
					b.SendMessage(SendMessageParams{Message: fmt.Sprintf("Failed to get %s's collection", event.ChatterUserName)})
					return
				}
				b.broadcast(server.OverlayEvent{Type: server.EventTypeCollectionDisplay, Data: blindbox.BlindBoxDisplayData{Username: event.ChatterUserName, Collection: slots, Config: cfg}})
				b.logger.Info("displaying collection", "user", event.ChatterUserName, "series", cfg.Series, "size", len(slots))
			},
		}
	}

	return cmds
}

File: web/src/lib/server/bot/commands.ts
import type { BlindBoxSeries } from '$lib/contracts/catalog';
import type {
  BlindBoxService,
  Logger,
  OverlayBus,
  StatsService,
} from '$lib/server/application/ports';
import { formatStats } from '$lib/server/domain/stats/format';

import { splitFields } from './text';
import type { Command, SendChat } from './types';

export interface CommandDependencies {
  series: readonly BlindBoxSeries[];
  stats: StatsService;
  blindBox: BlindBoxService;
  overlay: OverlayBus;
  logger: Logger;
  sendChat: SendChat;
}

export const createCommands = ({
  series,
  stats,
  blindBox,
  overlay,
  logger,
  sendChat,
}: CommandDependencies): ReadonlyMap<string, Command> => {
  const commands = new Map<string, Command>();

  commands.set('collections', async (_event, signal) => {
    let collections;
    try {
      collections = await blindBox.completed();
    } catch (error) {
      logger.error('failed to get completed collections', { error });

      return;
    }

    if (signal.aborted) {
      return;
    }

    await sendChat(
      'The following chatters have completed the below blind box collections:',
      signal,
    );
    for (const collection of collections) {
      if (signal.aborted) {
        return;
      }

      await sendChat(`${collection.seriesName}: ${collection.usernames}`, signal);
    }
  });

  commands.set('leaderboard', async (_event, signal) => {
    let rows;
    try {
      rows = await stats.leaderboard();
    } catch (error) {
      logger.error('failed to get leaderboard', { error });
      await sendChat('Failed to get leaderboard', signal);

      return;
    }

    if (signal.aborted) {
      return;
    }

    const message = rows.map((row) => `${row.emoji} ${row.username} (${row.value})`).join(' | ');
    await sendChat(message, signal);
  });

  commands.set('stats', async (event, signal) => {
    if (splitFields(event.text).length !== 1) {
      return;
    }

    let values;
    try {
      values = await stats.getOrCreate(event.chatterUserID, event.chatterUsername);
    } catch (error) {
      logger.error('failed to get stats', { error, user: event.chatterUsername });

      return;
    }

    if (signal.aborted) {
      return;
    }

    await sendChat(formatStats(event.chatterUsername, values), signal, event.messageID);
  });

  for (const config of series) {
    commands.set(config.series, async (event, signal) => {
      if (splitFields(event.text).length !== 1) {
        return;
      }

      let collection;
      try {
        collection = await blindBox.getCollection(event.chatterUserID, config.series);
      } catch (error) {
        logger.error('failed to get collection', { error, user: event.chatterUsername });
        await sendChat(`Failed to get ${event.chatterUsername}'s collection`, signal);

        return;
      }

      if (signal.aborted) {
        return;
      }

      overlay.publish({
        type: 'blindbox_display',
        username: event.chatterUsername,
        collection,
        config,
      });
      logger.info('displaying collection', {
        user: event.chatterUsername,
        series: config.series,
        size: collection.length,
      });
    });
  }

  return commands;
};

export const commandName = (text: string): string | undefined => {
  if (!text.startsWith('!')) {
    return undefined;
  }

  const fields = splitFields(text.toLowerCase());
  if (fields.length === 0) {
    return undefined;
  }

  const name = fields[0].startsWith('!') ? fields[0].slice(1) : fields[0];

  return name || undefined;
};
